user-org-keycloak-util/src/main/java/org/sunbird/sso/service/impl/KeyCloakServiceImpl.java

```diff
@@ -0,0 +1,486 @@
+package org.sunbird.sso.service.impl;
+
+import static java.util.Arrays.asList;
+
+import java.security.KeyFactory;
+import java.security.PublicKey;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.lang3.StringUtils;
+import org.keycloak.RSATokenVerifier;
+import org.keycloak.admin.client.Keycloak;
+import org.keycloak.admin.client.resource.UserResource;
+import org.keycloak.representations.AccessToken;
+import org.keycloak.representations.idm.CredentialRepresentation;
+import org.keycloak.representations.idm.UserRepresentation;
+import org.sunbird.exception.AuthorizationException;
+import org.sunbird.exception.ProjectCommonException;
+import org.sunbird.exception.message.IResponseMessage;
+import org.sunbird.exception.message.Localizer;
+import org.sunbird.exception.message.ResponseCode;
+import org.sunbird.sso.model.User;
+import org.sunbird.sso.service.ISSOService;
+import org.sunbird.sso.util.KeyCloakConnectionProvider;
+import org.sunbird.sso.util.SSOConstant;
+import org.sunbird.util.LoggerEnum;
+import org.sunbird.util.ProjectLogger;
+import org.sunbird.util.PropertiesCache;
+
+/**
+ * Single sign out service implementation with Key Cloak.
+ *
+ * @author Manzarul
+ */
+public class KeyCloakServiceImpl implements ISSOService {
+
+  private Keycloak keycloak = KeyCloakConnectionProvider.getConnection();
+  private static PublicKey SSO_PUBLIC_KEY = null;
+
+  public PublicKey getPublicKey() {
+    if (null == SSO_PUBLIC_KEY) {
```

Comment:
we already save this to cache , so first read from cache if null then read from keycloak

Reply:
SSO_PUBLIC_KEY is cache inside this method itself , so don't think it's really required . First time any ways we are not going to get this, second time there might be a chance for mutlithreading to overrider old value with new.